src/python_testing/TC_OPCREDS_3_9.py

```diff
@@ -0,0 +1,354 @@
+from chip.utils import CommissioningBuildingBlocks
+from ecdsa import NIST256p, VerifyingKey
+from ecdsa.keys import BadSignatureError
+from mobly import asserts
+
+
+class MatterCertParser:
+    SUBJECT_TAG = 6
+    SUBJECT_PUBLIC_KEY_TAG = 9
+    SUBJECT_FABRIC_ID_TAG = 21
+
+    def __init__(self, matter_cert_bytes: bytes):
+        self.parsed_tlv = TLVReader(matter_cert_bytes).get()["Any"]
+
+    def get_subject_names(self) -> dict[int, object]:
+        return {tag: value for tag, value in self.parsed_tlv[self.SUBJECT_TAG]}
+
+    def get_public_key_bytes(self) -> bytes:
+        public_key_bytes = self.parsed_tlv[self.SUBJECT_PUBLIC_KEY_TAG]
+        return public_key_bytes
+
+
+# From Matter spec src/crypto_primitives/crypto_primitives.py
+class MappingsV1(enum.IntEnum):
+    CHIP_CRYPTO_HASH_LEN_BITS = 256
+    CHIP_CRYPTO_HASH_LEN_BYTES = 32
+    CHIP_CRYPTO_HASH_BLOCK_LEN_BYTES = 64
+    CHIP_CRYPTO_GROUP_SIZE_BITS = 256
+    CHIP_CRYPTO_GROUP_SIZE_BYTES = 32
+    CHIP_CRYPTO_PUBLIC_KEY_SIZE_BYTES = (2 * CHIP_CRYPTO_GROUP_SIZE_BYTES) + 1
+    CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BITS = 128
+    CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES = 16
+    CHIP_CRYPTO_AEAD_MIC_LENGTH_BITS = 128
+    CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES = 16
+    CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES = 13
+
+
+# From Matter spec src/crypto_primitives/crypto_primitives.py
+def bytes_from_hex(hex: str) -> bytes:
+    """Converts any `hex` string representation including `01:ab:cd` to bytes
+
+    Handles any whitespace including newlines, which are all stripped.
+    """
+    return unhexlify("".join(hex.replace(":", "").split()))
+
+
+# From Matter spec src/crypto_primitives/vid_verify_payload_test_vector.py
+# ECDSA-with-SHA256 using NIST P-256
+FABRIC_BINDING_VERSION_1 = 0x01
+STATEMENT_VERSION_1 = 0x21
+SIGNATURE_SIZE = (2 * MappingsV1.CHIP_CRYPTO_GROUP_SIZE_BYTES)
+SKID_SIZE = 20
+VID_VERIFICATION_CLIENT_CHALLENGE_SIZE_BYTES = 32
+ATTESTATION_CHALLENGE_SIZE_BYTES = MappingsV1.CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES
+
+
+def verify_signature(public_key: bytes, message: bytes, signature: bytes) -> bool:
+    """Verify a `signature` on the given `message` using `public_key`. Returns True on success."""
+
+    verifying_key: VerifyingKey = VerifyingKey.from_string(public_key, curve=NIST256p)
+    assert verifying_key.curve == NIST256p
+
+    try:
+        return verifying_key.verify(signature, message, hashfunc=hashlib.sha256)
+    except BadSignatureError:
+        return False
+
+
+# From Matter spec src/crypto_primitives/vid_verify_payload_test_vector.py
+def generate_vendor_fabric_binding_message(
+        root_public_key_bytes: bytes,
+        fabric_id: int,
+        vendor_id: int) -> bytes:
+
+    assert len(root_public_key_bytes) == MappingsV1.CHIP_CRYPTO_PUBLIC_KEY_SIZE_BYTES
+    assert fabric_id > 0 and fabric_id <= 0xFFFF_FFFF_FFFF_FFFF
+    assert vendor_id > 0 and vendor_id <= 0xFFF4
+
+    fabric_id_bytes = fabric_id.to_bytes(length=8, byteorder='big')
```

Comment:
nit: maybe consider using something like https://construct.readthedocs.io/en/latest/intro.html#example to make these structures more readable for read/write.

Reply:
This is code that was verbatim from spec repos. Not updating.